pkg/kubectl/cmd/attach.go

```diff
@@ -111,146 +156,115 @@ func (*DefaultRemoteAttach) Attach(method string, url *url.URL, config *restclie
-
-	Attach        RemoteAttach
-	PodClient     coreclient.PodsGetter
-	GetPodTimeout time.Duration
-	Config        *restclient.Config
-}
-
 // Complete verifies command line arguments and loads data from the command environment
-func (p *AttachOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, argsIn []string) error {
-	if len(argsIn) == 0 {
-		return cmdutil.UsageErrorf(cmd, "at least 1 argument is required for attach")
-	}
-	if len(argsIn) > 2 {
-		return cmdutil.UsageErrorf(cmd, "expected POD, TYPE/NAME, or TYPE NAME, (at most 2 arguments) saw %d: %v", len(argsIn), argsIn)
-	}
-
-	namespace, _, err := f.ToRawKubeConfigLoader().Namespace()
+func (o *AttachOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, argsIn []string) error {
+	var err error
+	o.Namespace, _, err = f.ToRawKubeConfigLoader().Namespace()
 	if err != nil {
 		return err
 	}
 
-	p.GetPodTimeout, err = cmdutil.GetPodRunningTimeoutFlag(cmd)
-	if err != nil {
-		return cmdutil.UsageErrorf(cmd, err.Error())
-	}
-
-	builder := f.NewBuilder().
-		WithScheme(legacyscheme.Scheme).
-		NamespaceParam(namespace).DefaultNamespace()
-
-	switch len(argsIn) {
-	case 1:
-		builder.ResourceNames("pods", argsIn[0])
-	case 2:
-		builder.ResourceNames(argsIn[0], argsIn[1])
-	}
-
-	obj, err := builder.Do().Object()
-	if err != nil {
-		return err
-	}
+	o.AttachablePodFn = polymorphichelpers.AttachablePodForObjectFn
 
-	attachablePod, err := polymorphichelpers.AttachablePodForObjectFn(f, obj, p.GetPodTimeout)
+	o.GetPodTimeout, err = cmdutil.GetPodRunningTimeoutFlag(cmd)
 	if err != nil {
-		return err
+		return cmdutil.UsageErrorf(cmd, err.Error())
 	}
 
-	p.PodName = attachablePod.Name
-	p.Namespace = namespace
+	o.Builder = f.NewBuilder
+	o.Resources = argsIn
+	o.restClientGetter = f
 
 	fullCmdName := ""
 	cmdParent := cmd.Parent()
 	if cmdParent != nil {
 		fullCmdName = cmdParent.CommandPath()
 	}
 	if len(fullCmdName) > 0 && cmdutil.IsSiblingCommandExists(cmd, "describe") {
-		p.SuggestedCmdUsage = fmt.Sprintf("Use '%s describe pod/%s -n %s' to see all of the containers in this pod.", fullCmdName, p.PodName, p.Namespace)
+		o.SuggestedCmdUsage = fmt.Sprintf("Use '%s describe pod/%s -n %s' to see all of the containers in this pod.", fullCmdName, o.PodName, o.Namespace)
 	}
 
 	config, err := f.ToRESTConfig()
 	if err != nil {
 		return err
 	}
-	p.Config = config
+	o.Config = config
 
-	clientset, err := f.ClientSet()
-	if err != nil {
-		return err
-	}
-
-	p.PodClient = clientset.Core()
+	o.AttachFunc = defaultAttachFunc
 
-	if p.CommandName == "" {
-		p.CommandName = cmd.CommandPath()
+	if o.CommandName == "" {
+		o.CommandName = cmd.CommandPath()
 	}
 
 	return nil
 }
 
 // Validate checks that the provided attach options are specified.
-func (p *AttachOptions) Validate() error {
-	allErrs := []error{}
-	if len(p.PodName) == 0 {
-		allErrs = append(allErrs, errors.New("pod name must be specified"))
+func (o *AttachOptions) Validate(args []string) error {
+	if len(args) == 0 {
```

Comment:
You can use `o.Resources`, and don't pass `args` here at all. Maybe renaming then to `Args` would make it more obvious.